README
MyHD driver and utility package  --  http://myhd.sourceforge.net/
Copyright 2003-2007, Mike Bourgeous (some 2003, Mark Lehrer; see driver/README)
Licensed under the GPL.

doc/
	These files describe the known features of the chip
driver/
	Kernel module (no video capture yet - simply allows non-root to access
	the card and enables I2C)
tools/
	Use these to "play" with the card and figure out how stuff works.
tl880fb/
	Preliminary framebuffer driver.

There are step-by-step instructions on compiling the driver and running tests at
http://myhd.sourceforge.net/howtohelp.php.

For more information, please view the mailing list archives.  Also read all the
README files in this driver distribution (find . -name README).  Reading source
code is also good.

Remember that this driver has no video capture capability.  The kernel module
uses three devices per TL880 card in /dev/tl880 to allow all users to mmap the
memory regions of the TL880, or use the ioctl interface to control the card.
The tools are then used to upload the firmware, set external sync, turn on color
bars, display a mouse cursor, and tweak the OSD.  That's all that the driver can
do so far.  There is also I2C bus support.  For the latest known bugs, see the
web site and mailing list.

To test the I2C bus(es), you will need the lm_sensors package.  Load the
i2c-dev, i2c-algo-bit, and tl880.ko kernel modules, then run i2cdetect for each
bus displayed in the kernel message log (each card has a different number of
buses).  If you have any output beside all XX's, note the address and bus
number, and run i2cdump [bus] [address].  For example, the NXT2000 on the MyHD
has address 0x14 on bus 1, so i2cdetect 1 shows a response from address 0xa, and
i2cdump 1 0xa shows the registers of the NXT2000.


Still needed:

Someone with PCI DMA experience.

Someone who can figure out how to use WinDBG to peer into the Windows drivers.

Lots and lots of time.



vim: tw=80



ChangeLog
2008-06-27	Mike Bourgeous		<nitrogen at users.sourceforge.net>

	* Patched tools/dpc/videoout to work with set_gpio (should fix MDP-120
	  video output).

2007-11-01	Mike Bourgeous		<nitrogen at users.sourceforge.net>

	* Started making improvements to the interrupt handler, with the
	  eventual goal of separating the BH into individual functions and
	  allowing other drivers to register callbacks.

2007-09-14	Mike Bourgeous		<nitrogen at users.sourceforge.net>

	* Added a tl880_memcpy_swab32 function.

2007-09-12	Mike Bourgeous		<nitrogen at users.sourceforge.net>

	* Increased the minimum period size in the ALSA driver to make sure
	  that audio always works with mplayer and xine.

2007-09-09	Mike Bourgeous		<nitrogen at users.sourceforge.net>

	* Added boundary checks to tl880mem functions.
	* Changed the fb driver to toggle video passthrough when the
	  framebuffer is opened and closed.
	* Added a new sample rate setting function to tl880audio.c.

2007-09-08	Mike Bourgeous		<nitrogen at users.sourceforge.net>

	* Changed TL_ASSERT and CHECK_NULL* macros to include the source file
	  in the error message, and to use a configurable prefix string.
	* Decreased the resolution of the memory pool bitmap, as allocation of
	  the memory pool was failing.
	* Added tl880_memcpy().
	* Added tl880_card_count().
	* Put a lot of work into the ALSA driver.
	* Fixed the framebuffer driver to work with the current version of the
	  tl880 module.
	* Added a new tool, iocread4reg.
	* Made improvements to the ALSA driver's internal names.

2007-09-06	Mike Bourgeous		<nitrogen at users.sourceforge.net>

	* Made minor initialization tweaks.
	* Added genalloc init/cleanup to module init/cleanup.
	* Started tracking down genalloc init problems.

2007-09-05	Mike Bourgeous		<nitrogen at users.sourceforge.net>

	* Added a complete table of the supported sample rates of the MDP-110.
	* Made other progress on the audio front.
	* Added genalloc.c temporarily until its presence can be confirmed in
	  most distribution-compiled kernels.

2007-08-13	Mike Bourgeous		<nitrogen at users.sourceforge.net>
	
	* Finished the SetSamplingClock function.
	* Added a new file for frontend-related code.
	* Improved audio interrupt handling.
	* Added some DPC utility functions to be used by other parts of the
	  driver.

2007-08-12	Mike Bourgeous		<nitrogen at users.sourceforge.net>
	
	* Improved some documentation.
	* Started switching card memory allocation to use genalloc.
	* Minor improvements to the audio code comments.
	* Added initialization of new fields to tl880_create_dev.
	* Started interpreting the SetSamplingClock function.

2007-07-15	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Added some information about the audio/video clock chips to doc/*.

2007-04-23	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Made the driver 64-bit safe (at least, it compiles and works okay on
	  my Athlon 64 x2).
	* Updated all tools to use explicit 32-bit kernel types, making the
	  tools 64-bit compatible.
	* Removed unknown region support and added offset support to regload.

2007-03-29	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Added separate build rules to tools/Makefile for each tool, to speed
	  rebuilding of a single tool (it's no longer necessary to compile all
	  of the tools to compile just one).
	* Added a new tool, channel, to set the tuner frequency.
	* Added some minimal MSP init code.

2007-03-27	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Finished analyzing set_video_attribute VPX function (still needs a
	  bit of cleaning up), added some more VPX constants to tl880vpx.h.
	* Added a set_gpio ioctl and added an iocsetgpio tool.
	* Made some enhancements to the generic ioctl tool.
	* Changed tl880_init_dev to use detected card type instead of PCI IDs.

2007-03-26	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Added MDP-120/MDP-130 GPIO patch from Jason P. Matthews.
	* Tweaked the GPIO patch to use TL880 driver code conventions.
	* Updated the AUTHORS and THANKS files.
	* Chip documentation updates.
	* Corrected a transposition of lines in the initialization messages.

2007-03-20	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Several new DPC registers (partially) documented.
	* Fixed a bug in the interrupt bh where interrupts would be disabled
	  any time the bh (bottom half/deferred interrupt handler) was called.
	* Three new APU registers partially documented.
	* Further reduced verbosity of interrupt handler messages (use the
	  debug module parameter to increase interrupt messages).

2007-03-17	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Added memset tool.
	* Many improvements to tl880vpx.c.
	* Split VPX definitions from tl880.h to tl880vpx.h.
	* Reduced interrupt printk verbosity.
	* Other tweaks to the interrupt handler.
	* VPX improvements.
	* More chip info in tl880doc.txt.

2007-03-15	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Added interrupt counting to tl880int.c.
	* Added interrupt status variables to tl880.h for MCU interrupts.
	* No longer disable TSD interrupts when receiving one.
	* Added DMA memory offset registers to tl880doc.txt (thanks to the PCI
	  traces provided by Stephen Marchant).
	* Added report of DMA offset to /proc/tl880.

2007-02-18	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Implemented more APU initialization.
	* Small updates to the chip documentation.

2007-01-31	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Started working on more VPX initialization.

2007-01-29	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Added a status dump IOCTL.
	* Tweak implementation of /proc interface.

2007-01-28	Mike Bourgeous		<i_am_nitrogen at hotmail.com>

	* Check the (usually useless) return value of request_module.
	* Change i2c init to try to access more possible i2c buses on unknown
	  card types.
	* Add __init back to tl880_init() in tl880_kern.c.
	* Improve/alter some error messages.

driver/tl880.h
/* 
 * Driver for TL880-based cards (possibly also TL850)
 *
 * (c) 2003-2007 Mike Bourgeous <nitrogen at users.sourceforge.net>
 * (c) 2007 Jason P. Matthews
 *
 * $Log: tl880.h,v $
 * Revision 1.34  2008/06/28 02:12:39  nitrogen
 * Importing old changes.  See ChangeLog.
 *
 * Revision 1.33  2007/09/13 09:16:13  nitrogen
 * Audio improvements.  Framebuffer tweak.  Documentation improvements.
 *
 * Revision 1.32  2007/09/09 06:16:48  nitrogen
 * Started an ALSA driver.  New iocread4reg tool.  Driver enhancements.
 *
 */

#ifndef _TL880_H_
#define _TL880_H_

#ifdef __KERNEL__

/* Standard includes */
#include <linux/init.h>
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/version.h>
#include <linux/dma-mapping.h>
#include <linux/delay.h>
#include <linux/errno.h>
#include <linux/genalloc.h>
#include <linux/fs.h>
#include <linux/proc_fs.h>
#include <linux/cdev.h>
#include <linux/major.h>
#include <linux/slab.h>
#include <linux/mm.h>
#include <linux/poll.h>
#include <linux/pci.h>
#include <linux/signal.h>
#include <linux/byteorder/swabb.h>
#include <asm/io.h>
#include <linux/ioport.h>
#include <asm/pgtable.h>
#include <asm/page.h>
#include <linux/sched.h>
#include <asm/types.h>
#include <linux/types.h>
#include <linux/interrupt.h>
#include <asm/uaccess.h>
#include <linux/vmalloc.h>
#include <linux/videodev.h>
#include <linux/i2c.h>
#include <linux/i2c-algo-bit.h>
#include <media/tuner.h>
#include <media/v4l2-common.h>
#include <media/tvaudio.h>
//#include <../drivers/media/dvb/frontends/nxt200x.h>

/*** Driver definitions ***/

/* Compatibility with pre-2.6.19 kernels */
#ifndef IRQF_SHARED
#define PRE_2619
#define IRQF_SHARED SA_SHIRQ
#endif /* IRQF_SHARED */
#ifndef IRQF_DISABLED
#define IRQF_DISABLED SA_INTERRUPT
#endif /* IRQF_DISABLED */


/* Devices */
#define DEV_COUNT				64
#define DEV_MASK				0xfffc	/* Bits of minor number identifying specific TL880 card */
#define DEV_SHIFT				2	/* Bits to right-shift minor number to get card identifier */
#define FUNC_MASK				3	/* Card function identifying bits of minor number (i.e. mem/reg/unk) */

/* I2C */
#define I2C_CLIENTS_MAX				16
#define I2C_HW_B_TL880				0x1000dd

/* PCI */
#define PCI_VENDOR_ID_TERALOGIC			0x544c
#define PCI_DEVICE_ID_TERALOGIC_TL880		0x0350

#define PCI_SUBSYSTEM_VENDOR_ID_MIT		0x17a1
#define PCI_SUBSYSTEM_DEVICE_ID_MYHD		0x0001
#define PCI_SUBSYSTEM_DEVICE_ID_MYHD_MDP120	0x0120
#define PCI_SUBSYSTEM_DEVICE_ID_MYHD_MDP130	0x0130

#define PCI_SUBSYSTEM_VENDOR_ID_HAUPPAUGE	0x0070
#define PCI_SUBSYSTEM_DEVICE_ID_WINTV_HD	0x4600

#define PCI_SUBSYSTEM_VENDOR_ID_TELEMANN	0x1378
#define PCI_SUBSYSTEM_VENDOR_ID_ZERO		0x0000
#define PCI_SUBSYSTEM_DEVICE_ID_HIPIX		0x0000

/* We'll give each new manufacturer a 0x10 bump */
#define TL880_CARD_HIPIX		0x20

#define TL880_CARD_WINTV_HD		0x10

#define TL880_CARD_MYHD_MDP130		6
#define TL880_CARD_MYHD_MDP120		5
#define TL880_CARD_MYHD_MDP110		4
#define TL880_CARD_MYHD_MDP100		3
#define TL880_CARD_MYHD_MDP100A		2
#define TL880_CARD_JANUS		1
#define TL880_CARD_ZERO			0

/* Debug */
#ifndef TL_MODNAME
#define TL_MODNAME "tl880"
#endif /* TL_MODNAME */

#define TL_ASSERT(a) ( (a) ? (0) : (printk(KERN_ERR "%s: assertion '%s' failed in %s at %u of %s\n", TL_MODNAME, #a, __FUNCTION__, __LINE__, __FILE__), (1)) )
#define CHECK_NULL(a) ( (a) ? (0) : (printk(KERN_ERR "%s: NULL %s in %s at %u of %s\n", TL_MODNAME, #a, __FUNCTION__, __LINE__, __FILE__), (1)) )
#define CHECK_NULL_W(a) ( (a) ? (0) : (printk(KERN_WARNING "%s: NULL %s in %s at %u of %s\n", TL_MODNAME, #a, __FUNCTION__, __LINE__, __FILE__), (1)) )


/*** Driver types ***/
struct tl880_dev;

struct tl880_i2c_bus {
	int busid;	/* bus access type */
	int busno;	/* card bus number */
	struct i2c_adapter		i2c_adap;
	struct i2c_algo_bit_data	i2c_algo;
	struct tl880_dev		*dev;
	struct i2c_client		*i2c_clients[I2C_CLIENTS_MAX];
};

/* Interrupt callback descriptor */
typedef void (*int_func)(struct tl880_dev *, void *);
struct int_func_desc {
	int_func func;
	void *private_data;
	struct int_func_desc *next;
};

struct tl880_dev {
	/** Housekeeping stuff **/
	struct tl880_dev *next;			/* Linked list pointer */
	struct pci_dev *pcidev;			/* PCI device structure */
	u32 id;					/* Card number */
	u16 subsys_vendor_id;			/* Card vendor ID */
	u16 subsys_device_id;			/* Card device ID */
	u16 card_type;				/* Card type identifier */
	char name[64];				/* Card model name */
	struct semaphore *sem;			/* For reentry protection */

	/* Character device stuff */
	struct cdev *char_device;		/* Kernel character device handle */
	u32 major;				/* Major number for all tl880 devices (for convenience) */
	u32 minor;				/* First minor number for this card */


	/** Card's I/O and memory space **/
	void *memspace;				/* Mapped memory space */
	unsigned long memphys;			/* Physical memory address */
	unsigned long memlen;			/* Memory space length */
	void *regspace;				/* Mapped register space */
	unsigned long regphys;			/* Physical register address */
	unsigned long reglen;			/* Register space length */
	void *unkspace;				/* Mapped unknown space */
	unsigned long unkphys;			/* Physical unknown address */
	unsigned long unklen;			/* Unknown space length */

	struct gen_pool *pool;			/* Generic memory allocation pool */


	/** DMA buffer **/
	void *dmavirt;				/* System-side DMA buffer pointer */
	dma_addr_t dmaphys;			/* Bus-side DMA buffer pointer */


	/** I2C stuff **/
	int minbus;				/* Lower bus-type ID for this card */
	int maxbus;				/* Upper bus-type ID for this card */
	struct tl880_i2c_bus *i2cbuses;		/* Array of I2C buses on this card */


	/** Interrupt stuff **/
	struct tasklet_struct tasklet;		/* Tasklet for interrupt work */

	u32 irq;				/* Interrupt line */

	int elseint;				/* Do we share an interrupt? */

	u32 old_mask;				/* Previously used interrupt mask */
	u32 int_mask;				/* Global interrupt enable mask */
	u32 int_type;				/* Type of interrupt to service */
	u32 int_count;				/* Number of interrupts received */

	u32 vsc_mask;				/* VSC interrupt mask */
	u32 vsc_type;				/* Type of VSC interrupt received */
	u32 vsc_count;				/* Number of VSC interrupts received */
	
	u32 apu_mask;				/* APU interrupt mask */
	u32 apu_type;				/* Type of APU interrupt received */
	u32 apu_count;				/* Number of APU interrupts received */
	struct int_func_desc *apu_th_funcs;	/* Linked list of APU top-half callbacks */
	struct int_func_desc *apu_bh_funcs;	/* Linked list of APU bottom-half callbacks */

	u32 blt_mask;				/* BLT interrupt mask */
	u32 blt_type;				/* Type of BLT interrupt received */
	u32 blt_count;				/* Number of BLT interrupts received */

	u32 mce_mask;				/* MCE interrupt mask */
	u32 mce_type;				/* Type of MCE interrupt received */
	u32 mce_count;				/* Number of MCE interrupts received */

	u32 mcu_mask;				/* MCU interrupt mask */
	u32 mcu_type;				/* Type of MCU interrupt received */
	u32 mcu_count;				/* Number of MCU interrupts received */

	u32 vpip_mask;				/* VPIP interrupt mask */
	u32 vpip_type;				/* Type of VPIP interrupt received */
	u32 vpip_count;				/* Number of VPIP interrupts received */

	u32 hpip_mask;				/* HPIP interrupt mask */
	u32 hpip_type;				/* Type of HPIP interrupt received */
	u32 hpip_count;				/* Number of HPIP interrupts received */

	u32 dpc_mask;				/* DPC interrupt mask */
	u32 dpc_type;				/* Type of DPC interrupt received */
	u32 dpc_count;				/* Number of DPC interrupts received */

	u32 tsd_mask;				/* TSD interrupt mask */
	u32 tsd_type;				/* Type of TSD interrupt received */
	u32 tsd_count;				/* Number of TSD interrupts received */


	/** Audio (APU) State **/
	enum audio_mode_e { 
		ZERO = 0, ONE, TWO 
	} audio_mode;				/* Current audio mode */

	void *iec_buf;				/* Audio buffer in system RAM */
	u32 iau_base;				/* Base card memory address for audio unit */
	u32 iau_iba_reg;			/* Register used for base address */
	u32 iau_iea_reg;			/* Register used for end address */
	u32 iau_ira_reg;			/* Register used for read address */


	/** Audio Chip (MSP) State **/
	int msp_addr;				/* Set to 0 if no MSP */
	int msp_i2cbus;				/* Set to index into i2cbuses */
	int msp_i2cclient;			/* Set to index into i2cbuses.i2c_clients */

	
	/** Video State **/


	/** Video Chip (VPX) State **/
	int vpx_addr;				/* Set to 0 if no VPX */
	int vpx_i2cbus;				/* Set to index into i2cbuses */
	int vpx_i2cclient;			/* Set to index into i2cbuses.i2c_clients */
	enum video_standard_e {
		PAL_B = 0,
		PAL_G = 0,
		PAL_H = 0,
		PAL_I = 0,
		NTSC_M = 1,
		SECAM = 2,
		NTSC44 = 3,
		PAL_M = 4,
		PAL_N = 5,
		PAL_60 = 6,
		NTSC_COMB = 7,
		NTSC_COMP = 9,
		FIFTEEN = 15
	} vpx_video_standard;			/* Currently active video standard */
};

#else /* __KERNEL__ */

/*** Userspace-exclusive definitions and declarations ***/

/* Userspace includes */
#include <linux/types.h>

#endif /* __KERNEL__ */

/*** Userspace/kernel space shared definitions such as ioctls ***/

/* ioctl definitions are subject to change */

/* 
 * Read register IOCTL - register is read from parameter, then the value read
 * is written back to the parameter.  Single unsigned long.
 */
#define TL880IOCREADREG		_IOWR(0xdd, 0, __u32 *)

/*
 * Write register IOCTL - register is read from parameter[0], the value to
 * write is read from parameter[1].  Array [2] __u32.
 */
#define TL880IOCWRITEREG	_IOW(0xdd, 1, __u32 *)

/*
 * VIP state IOCTL - parameter is a pointer to VIP state int (0 off, 1 on, 2
 * special?)
 */
#define TL880IOCSETVIP		_IOW(0xdd, 2, __u32 *)

/*
 * Set cursor position IOCTL - cursor position is packed into __u32
 * as two shorts - i.e. (x & 0xffff) | (y << 16)
 */
#define TL880IOCSETCURSORPOS	_IOW(0xdd, 3, __u32 *)

/*
 * Set card GPIO state - pass two __u32s.
 * parameter[0] = gpio_line
 * parameter[1] = state
 */
#define TL880IOCSETGPIO		_IOW(0xdd, 4, __u32 *)


#ifdef __KERNEL__

/*** Kernel definitions which depend on the above ***/


/*** Driver variables ***/
extern int debug;
extern dev_t device_number;
extern int n_tl880s;
extern struct tl880_dev *tl880_list;

/*** Driver functions ***/
/* tl880util.c */
void set_bits(u32 *value, u32 reg, int high_bit, int low_bit, u32 setvalue);
struct tl880_dev *find_tl880(unsigned long tl880_id);
struct tl880_dev *find_tl880_pci(struct pci_dev *dev);
int tl880_card_count(void);

/* tl880i2c.c */
int tl880_init_i2c(struct tl880_dev *tl880dev);
void tl880_deinit_i2c(struct tl880_dev *tl880dev);
int tl880_call_i2c_clients(struct tl880_dev *tl880dev, u32 cmd, void *arg);
/* I2C access functions for features not handled by external chip drivers */
int tl880_i2c_read_byte(struct tl880_i2c_bus *i2cbus, u16 addr);
int tl880_i2c_write_byte(struct tl880_i2c_bus *i2cbus, u16 addr, u8 value);
int tl880_i2c_read_byte_data(struct tl880_i2c_bus *i2cbus, u16 addr, u8 command);
int tl880_i2c_write_byte_data(struct tl880_i2c_bus *i2cbus, u16 addr, u8 command, u8 value);
int tl880_i2c_read_word_data(struct tl880_i2c_bus *i2cbus, u16 addr, u8 command);
int tl880_i2c_write_word_data(struct tl880_i2c_bus *i2cbus, u16 addr, u8 command, u16 value);

/* tl880init.c */
void tl880_init_myhd(struct tl880_dev *tl880dev);
void tl880_init_wintv_hd(struct tl880_dev *tl880dev);
void tl880_init_hipix(struct tl880_dev *tl880dev);
void tl880_init_chip(struct tl880_dev *tl880dev);
void tl880_init_dev(struct tl880_dev *tl880dev);

void tl880_deinit_myhd(struct tl880_dev *tl880dev);
void tl880_deinit_wintv_hd(struct tl880_dev *tl880dev);
void tl880_deinit_hipix(struct tl880_dev *tl880dev);
void tl880_deinit_dev(struct tl880_dev *tl880dev);

/* tl880gpio.c */
u8 tl880_set_gpio(struct tl880_dev *tl880dev, u32 gpio_line, u8 state);
void tl880_write_gpio1_wintv_hd(struct tl880_dev *tl880dev, u8 state, u8 b, int c);

/* tl880int.c */
#ifdef PRE_2619 /* Pre-2.6.19 compatibility */
#warning Using old-style interrupt routine
irqreturn_t tl880_irq(int irq, void *dev_id, struct pt_regs *regs);
#else /* PRE_2619 */
irqreturn_t tl880_irq(int irq, void *dev_id);
#endif /* PRE_2619 */
void tl880_bh(unsigned long tl880_id);
void tl880_disable_interrupts(struct tl880_dev *tl880dev);

int tl880_add_int_callback(struct tl880_dev *tl880dev, struct int_func_desc **list, int_func func, void *private_data);
void tl880_remove_int_callback(struct tl880_dev *tl880dev, struct int_func_desc **list, int_func func);

int tl880_register_apu_th(struct tl880_dev *tl880dev, int_func func, void *private_data);
void tl880_unregister_apu_th(struct tl880_dev *tl880dev, int_func func);

/* tl880vip.c */
void tl880_set_vip(struct tl880_dev *tl880dev, unsigned long vip_mode);

/* tl880input.c */
void tl880_set_video_source(struct tl880_dev *tl880dev, u8 a, u8 b);
void tl880_set_ntsc_input(struct tl880_dev *tl880dev, int input);

/* tl880audio.c */
void tl880_disable_apu(struct tl880_dev *tl880dev);
void tl880_apu_init_iau_reg(struct tl880_dev *tl880dev);
void tl880_apu_init_ioc(struct tl880_dev *tl880dev);
void tl880_apu_start_ioc(struct tl880_dev *tl880dev);
void tl880_apu_stop_ioc(struct tl880_dev *tl880dev);

void tl880_init_hardware_audio(struct tl880_dev *tl880dev, enum audio_mode_e audio_mode);
void tl880_deinit_hardware_audio(struct tl880_dev *tl880dev);

void tl880_init_ntsc_audio(struct tl880_dev *tl880dev);
void tl880_ntsc_audio_dpc(struct tl880_dev *tl880dev);
void tl880_set_ntsc_audio_clock(struct tl880_dev *tl880dev);

void tl880_set_sampling_clock(struct tl880_dev *tl880dev, int rate); /* should be an enum */

int tl880_audio_set_crossfade(struct tl880_dev *tl880dev, s32 arg_0, s32 arg_4, u8 arg_8, u8 arg_C,
							  u8 arg_10, u8 arg_14, u8 arg_18, u8 arg_1C);

int tl880_set_samplerate(struct tl880_dev *tl880dev, int rate);

/* tl880demux.c */
unsigned long tl880_demux_init(struct tl880_dev *tl880dev);

/* tl880dma.c */
u32 tl880_aux_dma_allocate(void);
void tl880_aux_dma_free(u32 dma);
extern unsigned long dma_bitmask;

int tl880_dma_vop_on(struct tl880_dev *tl880dev);

/* tl880proc.c */
int tl880_create_proc_entry(void);
void tl880_remove_proc_entry(void);

/* tl880mem.c */
u32 tl880_read_memory(struct tl880_dev *tl880dev, u32 mem);
void tl880_write_memory(struct tl880_dev *tl880dev, u32 mem, u32 value);
u32 tl880_read_membits(struct tl880_dev *tl880dev, u32 mem, int high_bit, int low_bit);
void tl880_write_membits(struct tl880_dev *tl880dev, u32 mem, int high_bit, int low_bit, u32 value);
void tl880_memcpy(struct tl880_dev *tl880dev, void *src, u32 addr, size_t length);
void tl880_memcpy_swab32(struct tl880_dev *tl880dev, void *src, u32 addr, size_t length);
void tl880_memcpy_swahw32(struct tl880_dev *tl880dev, void *src, u32 addr, size_t length);
void tl880_clear_sdram(struct tl880_dev *tl880dev, u32 start_addr, u32 end_addr, u32 value);

int tl880_init_memory(struct tl880_dev *tl880dev);
void tl880_deinit_memory(struct tl880_dev *tl880dev);
unsigned long tl880_alloc_memory(struct tl880_dev *tl880dev, size_t bytes);
void tl880_free_memory(struct tl880_dev *tl880dev, unsigned long addr, size_t bytes);


#endif /* __KERNEL__ */


/*** Driver local includes ***/

#include "tl880reg.h"
#include "tl880dpc.h"
#include "tl880osd.h"
#include "tl880vpx.h"
#include "tl880msp.h"

#endif /* _TL880_H_ */

